storage/block/block.go

```diff
@@ -42,31 +42,49 @@ var (
 type dbBlock struct {
 	sync.RWMutex
 
-	opts          Options
-	ctx           context.Context
+	// NB(r): we cache this instance's stream context finalizer
+	// function to avoid allocating a reference to the fn
+	// every time a read occurs
+	onStreamDoneFinalizer context.Finalizer
+
+	opts Options
+	ctx  context.Context
+
 	startUnixNano int64
 	segment       ts.Segment
 	length        int
 	checksum      uint32
 
 	accessedUnixNano int64
 
-	mergeTarget DatabaseBlock
-
 	retriever  DatabaseShardBlockRetriever
 	retrieveID ts.ID
 
+	// For use in a list of blocks, notably for the wired list. The
+	// doubly linked list pointers are on the block themselves to avoid
+	// creating a heap allocated struct per block.
+	// NB(r): Beware that the list controls the synchronization of these
+	// fields to avoid unnecessary lock contention.
+	// They should never be read or mutated by the block itself.
+	next *dbBlock
```

Comment:
Yeah so this actually can't be done in Go as when you follow the pointer to the next element, there's no way to actually get a reference to the enclosing struct.  e.g.:

```go
type list struct {
  root listElem
}

type listElem struct {
  next *listElem

  // you could create an `interface{}` type field here that lets you reference back to the enclosing struct.. however that would cause slow and gross cast back to the enclosing struct
}

type foo struct {
  listElem

  x int
  // other fields 
}

func useList() {
  // create list
  l := &list{}
  l.root.next = &l.root.listElem

  // insert
  x, y, z := &foo{x: 1}, &foo{x: 2}, &foo{x: 3}
  l.root.next = &x.listElem
  x.next = &y.listElem
  y.next = &z.listElem

  // try to iterate and access the "foo" fields
  for elem := l.root.next; elem.next != &l.root.listElem; elem = elem.next {
    // only have a reference to *listElem now
  }
}
```